core/runtime/src/main/java/io/quarkus/runtime/shutdown/ShutdownRecorder.java

```diff
@@ -16,41 +15,36 @@ public class ShutdownRecorder {
     private static final Logger log = Logger.getLogger(ShutdownRecorder.class);
 
     private static volatile List<ShutdownListener> shutdownListeners;
-    private static volatile Optional<Duration> waitTime;
-
-    final ShutdownConfig shutdownConfig;
+    private static volatile ShutdownConfig shutdownConfig;
 
     public ShutdownRecorder(ShutdownConfig shutdownConfig) {
-        this.shutdownConfig = shutdownConfig;
+        ShutdownRecorder.shutdownConfig = shutdownConfig;
     }
 
     public void setListeners(List<ShutdownListener> listeners) {
-        shutdownListeners = listeners;
-        waitTime = shutdownConfig.timeout;
+        shutdownListeners = Optional.ofNullable(listeners).orElseGet(List::of);
     }
 
     public static void runShutdown() {
-        if (shutdownListeners == null) {
-            return;
-        }
-        log.debug("Attempting to gracefully shutdown.");
         try {
             CountDownLatch preShutdown = new CountDownLatch(shutdownListeners.size());
             for (ShutdownListener i : shutdownListeners) {
                 i.preShutdown(new LatchShutdownNotification(preShutdown));
             }
-
             preShutdown.await();
+
+            if (shutdownConfig.isDelaySet()) {
+                Thread.sleep(shutdownConfig.delay.get().toMillis());
```

Comment:
Maybe interruption here should be interpreted as "exit pre-shutdown, go to shutdown immediately"?

Reply:
Sure can do. Should we also re-interrupt the current thread (`Thread.currentThread().interrupt()`)?

Reply:
Done.